hudi-common/src/main/java/org/apache/hudi/internal/schema/utils/AvroSchemaEvolutionUtils.java

```diff
@@ -109,40 +116,129 @@ public static InternalSchema reconcileSchema(Schema incomingSchema, InternalSche
+    TableChanges.ColumnUpdateChange schemaChange =
+        reduce(reconciliationTargetColumnNamePairs, TableChanges.ColumnUpdateChange.get(sourceInternalSchema, true),
+            (change, sourceTargetPair) -> change.renameColumn(sourceTargetPair.getLeft(), sourceTargetPair.getRight()));
+
+    return convert(SchemaChangeUtils.applyTableChanges2Schema(sourceInternalSchema, schemaChange), sourceSchema.getName());
+  }
+
+  /**
+   * Resolves fields in the {@code source} schema matching {@code target} schema, even though
+   * their letter casing might be mismatched. Returns fully-qualified names pairs for the fields from
+   * the {@code source} schema for which, there is correspondent field in the {@code target} schema
+   * that only differs in its letters case.
+   *
+   * For example, provided two schemas {@code A} and {@code B}, such that:
+   * {@code struct A { fieldA: ... }}, while {@code struct B { FieldA: ... }}, this method will
+   * produce following list: {@code Map("fieldA" -> "FieldA")}
+   *
+   * NOTE: This method doesn't perform validation whether types of the correspondent fields are
+   *       conformant, and simply applies a case-insensitive resolution logic
+   */
+  public static Map<String, String> resolveCaseMismatchingFieldNamePairs(Schema sourceSchema,
+                                                                                Schema targetSchema) {
+    if (sourceSchema.getFields().isEmpty() || targetSchema.getFields().isEmpty()) {
+      return Collections.emptyMap();
+    }
+
+    InternalSchema sourceInternalSchema = convert(sourceSchema);
+    InternalSchema targetInternalSchema = convert(targetSchema);
+
+    return resolveCaseMismatchingFieldNamePairs(sourceInternalSchema, targetInternalSchema)
+        .stream()
+        .collect(Collectors.toMap(Pair::getLeft, Pair::getRight));
+  }
+
+  private static List<Pair<String, String>> resolveCaseMismatchingFieldNamePairs(InternalSchema sourceSchema,
+                                                                                 InternalSchema targetSchema) {
+    List<String> colNamesSourceSchema = sourceSchema.getAllColsFullName();
+
+    // To reconcile field names of the source schema we enumerate all possible fields (including
+    // nested ones) and produce a mapping convert them to lower-case to correlate across
+    // source and target schemas
+    Map<String, String> colNamesMapTargetSchema =
+        targetSchema.getAllColsFullName().stream()
+            .collect(Collectors.toMap(String::toLowerCase, Function.identity()));
+
+    return colNamesSourceSchema.stream()
+        .map(sourceFieldName -> {
+          // NOTE: Here we're only reconciling the fields that differ only in casing of their
+          //       corresponding names (therefore a) fields have to be present in both schemas and
+          //       b) they have to differ)
+          String lowerCaseSourceFieldName = sourceFieldName.toLowerCase(Locale.ROOT);
```

Comment:
we should do rename from the bottom.
eg:   sourceCol : a.b.c  targetCol: A.B.C
we shoud do rename as fllow
1) rename a.b.c -> C
2) rename a.b -> B
3) rename a -> A

Reply:
```
    return colNamesSourceSchema.stream()
        .map(sourceFieldName -> {
          // NOTE: Here we're only reconciling the fields that differ only in casing of their
          //       corresponding names (therefore a) fields have to be present in both schemas and
          //       b) they have to differ)
          String lowerCaseSourceFieldName = sourceFieldName.toLowerCase(Locale.ROOT);
          String targetFieldName = colNamesMapTargetSchema.get(lowerCaseSourceFieldName);
          if (targetFieldName == null) {
            return null;
          }
          String rawNameFromSourceField = getRawName(sourceFieldName);
          String rawNameFromTargetField = getRawName(targetFieldName);

          if (!Objects.equals(rawNameFromSourceField, rawNameFromTargetField)) {
            return Pair.of(sourceFieldName, rawNameFromTargetField);
          } else {
            return null;
          }
        })
        .filter(Objects::nonNull)
        .sorted((o1, o2)  -> o2.getLeft().split("\\.").length - o1.getLeft().split("\\.").length)
        .collect(Collectors.toList());
  }

  private static String getRawName(String colName) {
    int splitPoint = colName.lastIndexOf(".");
    return splitPoint > 0 ? colName.substring(splitPoint + 1) : colName;
  }
```